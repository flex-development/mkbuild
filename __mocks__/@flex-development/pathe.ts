/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
