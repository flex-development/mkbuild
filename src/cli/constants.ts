/**
 * @file CLI - Constants
 * @module mkbuild/cli/constants
 */

import pkg from '#pkg' assert { type: 'json' }

/**
 * Array containing choices for boolean flags.
 *
 * @internal
 *
 * @const {string[]} CHOICES_BOOLEAN
 */
export const CHOICES_BOOLEAN: string[] = ['false', 'true', '0', '1']

/**
 * CLI application name.
 *
 * @internal
 *
 * @const {string} CLI_NAME
 */
export const CLI_NAME: string = pkg.name.replace(/.*\//, '')
