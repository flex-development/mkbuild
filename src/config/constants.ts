/**
 * @file Config - Constants
 * @module mkbuild/config/constants
 */

/**
 * Default ignore patterns.
 *
 * @see https://github.com/mrmlnc/fast-glob#ignore
 *
 * @const {string[]} IGNORE_PATTERNS
 */
export const IGNORE_PATTERNS: string[] = [
  '**/.DS_Store',
  '**/.npmignore',
  '**/.yarnignore',
  '**/**.mdx',
  '**/**.stories.{j,t}sx',
  '**/{__mocks__,__snapshots__,__tests__}/**'
]

/**
 * Module extensions checked when attempting to resolve path aliases and module
 * specifiers.
 *
 * @const {string[]} MODULE_EXTENSIONS
 */
export const MODULE_EXTENSIONS: string[] = [
  '.cjs',
  '.cts',
  '.js',
  '.json',
  '.jsx',
  '.mdx',
  '.mjs',
  '.mts',
  '.ts',
  '.tsx'
]

/**
 * `require` and `require.resolve` statement regex.
 *
 * @todo prevent matching non-statements (e.g. comments, arbitrary strings)
 *
 * @const {RegExp} REQUIRE_REGEX
 */
export const REQUIRE_REGEX: RegExp =
  /(?:require\(|require\.resolve\()["']([^"']*)["']\)/g
