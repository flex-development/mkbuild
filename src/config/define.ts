/**
 * @file Config - defineBuildConfig
 * @module mkbuild/config/defineBuildConfig
 */

import type { Config } from '#src/interfaces'

/**
 * Helper for defining a build configuration object.
 *
 * @param {Config} [config={}] - Build configuration options
 * @return {Config} `config`
 */
const defineBuildConfig = (config: Config = {}): Config => config

export default defineBuildConfig
