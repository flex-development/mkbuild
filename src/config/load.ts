/**
 * @file Config - loadBuildConfig
 * @module mkbuild/config/loadBuildConfig
 */

import type { Config } from '#src/interfaces'
import { cosmiconfig } from 'cosmiconfig'
import es from './loader-es'

/**
 * Searches for a build configuration file.
 *
 * If found, a config object will be returned. The object will contain options
 * from the config file. If not found, an empty object will be returned.
 *
 * @see https://github.com/davidtheclark/cosmiconfig#explorersearch
 *
 * @async
 *
 * @param {string} [location] - Directory to search
 * @return {Config} Build configuration options
 */
const loadBuildConfig = async (location: string): Promise<Config> => {
  /**
   * Module name.
   *
   * @const {string} name
   */
  const name: string = 'build'

  // get search function
  const { search } = cosmiconfig(name, {
    loaders: { '.cts': es, '.js': es, '.mjs': es, '.mts': es, '.ts': es },
    searchPlaces: [
      '#.config.cjs',
      '#.config.cts',
      '#.config.js',
      '#.config.json',
      '#.config.mjs',
      '#.config.mts',
      '#.config.ts'
    ].map(place => place.replace('#', name)),
    stopDir: location
  })

  return ((await search(location))?.config as Config | null) ?? {}
}

export default loadBuildConfig
