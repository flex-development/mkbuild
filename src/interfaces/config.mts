/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystem, Task } from '@flex-development/mkbuild'

/**
 * Build configuration.
 *
 * @see {@linkcode Task}
 *
 * @extends {Task}
 */
interface Config extends Task {
  /**
   * File system API.
   *
   * @see {@linkcode FileSystem}
   */
  fs?: Partial<FileSystem> | null | undefined

  /**
   * Build tasks.
   *
   * Each build task will be merged with the base task. Top-level lists (arrays
   * and sets) are concatted and deduped. Plain objects are deep merged.
   *
   * > ðŸ‘‰ **Note**: If empty, `null`, or `undefined`, a single build task will
   * > be inferred from the base task.
   *
   * @see {@linkcode Task}
   */
  tasks?: Task[] | null | undefined
}

export type { Config as default }
