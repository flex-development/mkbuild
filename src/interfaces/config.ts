/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystemAdapter } from '#src/types'
import type * as esbuild from 'esbuild'
import type Options from './options'
import type Task from './task'

/**
 * Build configuration options.
 *
 * @extends {Options}
 */
interface Config extends Options {
  /**
   * Load build config from config file.
   *
   * @default true
   */
  configfile?: boolean

  /**
   * Build entries.
   *
   * **Note**: If empty or `undefined`, a single build entry will be inferred
   * from the remaining set of configuration options.
   *
   * @default []
   */
  entries?: Partial<Omit<Task, 'write'>>[]

  /**
   * Custom implementations of file system methods.
   *
   * @see {@linkcode FileSystemAdapter}
   * @see https://nodejs.org/api/fs.html
   *
   * @default fsa
   */
  fs?: FileSystemAdapter

  /**
   * Serve files.
   *
   * @see {@linkcode esbuild.ServeOptions}
   * @see https://esbuild.github.io/api/#serve
   *
   * @default false
   */
  serve?: esbuild.ServeOptions | boolean

  /**
   * Watch files.
   *
   * @default false
   */
  watch?: boolean

  /**
   * Write build results.
   *
   * @default false
   */
  write?: boolean
}

export type { Config as default }
