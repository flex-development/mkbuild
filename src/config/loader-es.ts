/**
 * @file Config - esLoader
 * @module mkbuild/config/esLoader
 */

import type { Config } from '#src/interfaces'
import tsconfigPaths from '#src/plugins/tsconfig-paths/plugin'
import { build } from 'esbuild'
import { resolveImports, toDataURL } from 'mlly'
import { RESOLVE_EXTENSIONS } from './constants'

/**
 * Loads an [ESM][1] (`*.js`, `*.mjs`) or TypeScript (`*.cts`, `*.mts`, `*.ts`)
 * build config file.
 *
 * [1]: https://nodejs.org/api/esm.html
 *
 * @see https://github.com/davidtheclark/cosmiconfig#loaders
 *
 * @async
 *
 * @param {string} path - Config file path
 * @param {string} content - Config file content
 * @return {Promise<Config>} Build configuration options
 */
const esLoader = async (path: string, content: string): Promise<Config> => {
  const {
    outputFiles: [output]
  } = await build({
    entryPoints: [path],
    format: 'esm',
    loader: {
      '.cts': 'ts',
      '.js': 'js',
      '.mjs': 'js',
      '.mts': 'ts',
      '.ts': 'ts'
    },
    metafile: true,
    outdir: '<stdout>',
    plugins: [tsconfigPaths()],
    write: false
  })

  // resolve imports in content to evaluate module
  content = await resolveImports(output!.text, {
    extensions: RESOLVE_EXTENSIONS,
    url: path
  })

  return ((await import(toDataURL(content))) as { default: Config }).default
}

export default esLoader
