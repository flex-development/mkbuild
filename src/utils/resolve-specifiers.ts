/**
 * @file Utilities - resolveSpecifiers
 * @module mkbuild/utils/resolveSpecifiers
 */

import { MODULE_EXTENSIONS } from '#src/config/constants'
import type { Entry } from '#src/interfaces'
import type { Metafile, OutputFile } from 'esbuild'
import * as pathe from 'pathe'
import extractStatements from './extract-statements'
import resolveSpecifier from './resolve-specifier'

/**
 * Adds file extensions to relative specifiers in `output` file content.
 *
 * @see {@link resolveSpecifier}
 *
 * @async
 *
 * @param {OutputFile} output - Output file object
 * @param {Uint8Array} output.contents - Output file content
 * @param {string} output.path - Absolute path to output file
 * @param {string} output.text - `output.contents` as string
 * @param {Pick<Metafile, 'outputs'>} metadata - Build metadata
 * @param {Entry['format']} [format] - Output file format
 * @param {Entry['ext']} [ext] - Output file extension
 * @param {string} [cwd=process.cwd()] - Current working directory
 * @param {string[]} [extensions=MODULE_EXTENSIONS] - Extensions to probe for
 * @return {Promise<OutputFile>} `output` with specifiers fully resolved
 */
const resolveSpecifiers = async (
  output: OutputFile,
  metadata: Pick<Metafile, 'outputs'>,
  format: Entry['format'],
  ext: Entry['ext'],
  cwd: string = process.cwd(),
  extensions: string[] = MODULE_EXTENSIONS
): Promise<OutputFile> => {
  /**
   * Relative path to output file.
   *
   * **Note**: Relative to {@link cwd}.
   *
   * @const {string} outfile
   */
  const outfile: string = output.path.replace(cwd + '/', '')

  // do nothing if output does not have metadata
  if (!metadata.outputs[outfile]) return output

  // do nothing if output does not have entry point
  if (!metadata.outputs[outfile]!.entryPoint) return output

  /**
   * {@link output.text} copy.
   *
   * @var {string} text
   */
  let text: string = output.text

  for (const statement of extractStatements(output.text, format)) {
    // do nothing if missing module specifier
    if (!statement.specifier) continue

    /**
     * {@link statement.code} before specifier resolution.
     *
     * @const {string} code
     */
    const code: string = statement.code

    // add extension to module specifier
    await resolveSpecifier(
      statement,
      pathe.resolve(cwd, metadata.outputs[outfile]!.entryPoint!),
      format,
      ext,
      extensions
    )

    // replace code in text
    text = text.replace(code, statement.code)
  }

  return { ...output, contents: new Uint8Array(Buffer.from(text)), text }
}

export default resolveSpecifiers
