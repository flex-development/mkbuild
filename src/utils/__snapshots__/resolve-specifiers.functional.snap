// Vitest Snapshot v1

exports[`functional:utils/resolveSpecifiers > resolve > should add file extensions to relative specifiers 1`] = `
"import type { Metafile, OutputFile } from 'esbuild'
import * as pathe from 'pathe'
import { MODULE_EXTENSIONS } from '../config/constants.mjs'
import type { Entry } from '../interfaces/index.mjs'
import extractStatements from './extract-statements.mjs'
import resolveSpecifier from './resolve-specifier.mjs'"
`;
